### DISCUSSION TITLE
Support for Table Comments

### DISCUSSION THREAD

**Discussion by TonyGravagno** [araera111/mysql-to-zod#34] [Ideas]
The Create Table command supports a table option to save a Comment for the table itself.

```sql
CREATE TABLE `airport` (
   ...
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
) ENGINE=InnoDB DEFAULT CHARSET=latin1 COMMENT='International Commercial Airports'
```

I'd like to add an enhancement later to save this comment into the schema:

```ts
// [table:airport] : International Commercial Airports
export const AirportSchema = z.object({
  id: z.number(), ...
```

Not only is this a simple reference back to the table, but with a specific pattern ( "[table:*]" ) it's easily searchable in a project. In the same enhancement or maybe a follow-up, I can make the prefix customizable.

**Comment by araera111** [comment]
I parsed a CREATE TABLE statement with node-sql-parser.
The following statement.

```sql
CREATE TABLE `todo` (
  `id` int NOT NULL AUTO_INCREMENT,
  `task` varchar(255) NOT NULL,
  `due_date` date DEFAULT NULL,
  `created_at` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
  `updated_at` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `comment` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci COMMENT='MY_TODO_LIST'
```

The result of the parse is as follows.
It is long, but see the ```table_options```. The last part of JSON. 

```JSON
{
  "type": "create",
  "keyword": "table",
  "temporary": null,
  "if_not_exists": null,
  "table": [
    {
      "db": null,
      "table": "todo"
    }
  ],
  "ignore_replace": null,
  "as": null,
  "query_expr": null,
  "create_definitions": [
    {
      "column": {
        "type": "column_ref",
        "table": null,
        "column": "id"
      },
      "definition": {
        "dataType": "INT",
        "suffix": []
      },
      "resource": "column",
      "nullable": {
        "type": "not null",
        "value": "not null"
      },
      "auto_increment": "auto_increment"
    },
    {
      "column": {
        "type": "column_ref",
        "table": null,
        "column": "task"
      },
      "definition": {
        "dataType": "VARCHAR",
        "length": 255
      },
      "resource": "column",
      "nullable": {
        "type": "not null",
        "value": "not null"
      }
    },
    {
      "column": {
        "type": "column_ref",
        "table": null,
        "column": "due_date"
      },
      "definition": {
        "dataType": "DATE"
      },
      "resource": "column",
      "default_val": {
        "type": "default",
        "value": {
          "type": "null",
          "value": null
        }
      }
    },
    {
      "column": {
        "type": "column_ref",
        "table": null,
        "column": "created_at"
      },
      "definition": {
        "dataType": "TIMESTAMP"
      },
      "resource": "column",
      "nullable": {
        "type": "null",
        "value": "null"
      },
      "default_val": {
        "type": "default",
        "value": {
          "type": "function",
          "name": "CURRENT_TIMESTAMP",
          "over": null
        }
      }
    },
    {
      "column": {
        "type": "column_ref",
        "table": null,
        "column": "updated_at"
      },
      "definition": {
        "dataType": "TIMESTAMP"
      },
      "resource": "column",
      "nullable": {
        "type": "null",
        "value": "null"
      },
      "default_val": {
        "type": "default",
        "value": {
          "type": "function",
          "name": "CURRENT_TIMESTAMP",
          "over": {
            "type": "on update",
            "keyword": "CURRENT_TIMESTAMP"
          }
        }
      }
    },
    {
      "column": {
        "type": "column_ref",
        "table": null,
        "column": "comment"
      },
      "definition": {
        "dataType": "VARCHAR",
        "length": 255
      },
      "resource": "column",
      "default_val": {
        "type": "default",
        "value": {
          "type": "null",
          "value": null
        }
      }
    },
    {
      "constraint": null,
      "definition": [
        {
          "type": "column_ref",
          "table": null,
          "column": "id"
        }
      ],
      "constraint_type": "primary key",
      "keyword": null,
      "index_type": null,
      "resource": "constraint",
      "index_options": null
    }
  ],
  "table_options": [
    {
      "keyword": "engine",
      "symbol": "=",
      "value": "INNODB"
    },
    {
      "keyword": "default charset",
      "symbol": "=",
      "value": "utf8mb4"
    },
    {
      "keyword": "collate",
      "symbol": "=",
      "value": "utf8mb4_0900_ai_ci"
    },
    {
      "keyword": "comment",
      "symbol": "=",
      "value": "'MY_TODO_LIST'"
    }
  ]
}
```

If you have ```"keyword": "comment"``` here, it would be possible to comment the value before the schema.
It would not be that difficult to implement.

This could be made optional, but I feel that if the table has a comment, it might be ok to automaticaly add the comment.
If it is optional, what kind of item should it be?
```hideTableComment: boolean```
or
```addTableComment: boolean```

Any suggestions?

**Comment by TonyGravagno** [comment]
I can't imagine anyone objecting to a comment. I would default the comment IN and have an option to exclude it out.

_Your use of node-sql-parser compelled me to switch from using MySQL-Zod. While a good project, it does not use CREATE TABLE as a source, and therefore does not have metadata like this. Rather than invest more time in my fork of that project, I decided to spend time with you here. Thank you for your coding choices._

**About options...**

When working on new features I have found that after first implementing them as Boolean, I often need to migrate to an object or array. So sometimes I will use `"feature": { "active": true }` and then add more nuanced options under that feature category as-necessary. Even if there are just two options related to a given feature, this structure keeps the root node of the options from getting too cluttered. And this allows for re-using option keywords for consistency. For example, rather than `useFeature1` and `useFeature2`, there might be two options, feature1 and feature2 and both of them have a 'use' flag. This is not a "rule" - it's more about elegance...

I haven't thought about this much but here is an idea for an option structure for comments. This is suitable for this topic of Table Comments, as well as #29 for Field Comments.

```ts
const options = {
  isAddType: true,
  // ...
  comments: {
    table: {
      active: true,
      format: '// [table:!name] : !text',
    },
    fields: {
      active: true,
      format: '// !name : !text',
    },
  },
}
```

The format can be anything the user/developer wants, and this utility will substitute run-time values for keywords '!name' and '!text'. Someone might want to use a different comment format like `//* ...\n... */` , so the toZod code won't insert a default `//` prefix, it will insert the entire format string after replacements.

What do you think?

**Comment by araera111** [comment]
I think your proposed option is reasonable. First, I implemented it for table comments. I wrote various tests so it may be working correctly.

I will add more options to the README.
I need to make the README more clear.

I will close my comment, but if you have any other comments, please re-open it.
